### DISCUSSION TITLE
date and time events

### DISCUSSION THREAD

**Discussion by imed-ghomari** [Taitava/obsidian-shellcommands#167] [Ideas]
Hi, is there a way to execute shell commands on a specific date and time ? Or is this type of event not yet implemented ?

**Comment by Taitava** [comment]
Hi, thank you for your question! 🙂

It's not possible at the moment, at least not easily. A not so easy to use workaround for now would be something like this:
- Use the [_Every n seconds_ event](https://publish.obsidian.md/shellcommands/Events/Every+n+seconds). Configure it to run e.g. every 60 seconds.
- Then in your shell command, have some kind of `if` condition to check if the time/date is right: `if [ {{date:HH:mm}} -eq "12:00" ]; then YOURCOMMANDHERE; fi` (This assumes your shell is Bash, but if you are using Windows, please let me know and I can post something for CMD or PowerShell)

Couple of links if you need more information on how to write if statements on Bash:
- https://superuser.com/questions/1169664/bash-if-on-single-line
- https://ryanstutorials.net/bash-scripting-tutorial/bash-if-statements.php

And in case you need, here is information about [all the format options for the `{{date:}}` variable](https://momentjs.com/docs/#/displaying/format/).

# A better solution for the future

I'm certainly interested in developing an event for executing commands based on a configurable schedule. I'm just not yet sure what scheduling format would be used, **and I'd like to hear opinions**.

The first thing that comes to my mind is the [Unix cron format](https://www.ibm.com/docs/en/db2oc?topic=task-unix-cron-format). It's a short textual format. I'll write just a few examples here:

- `* * * * *`: Each of these stars are a value, configuring in the respective order: _minute_, _hour_, _day of month_, _month_, _day of week_. Star means that the value does not matter, and now that all values are a star, the shell command would be executed every minute.
- `0 * * * *`: Execute the shell command every time the _minute_ is `0`, i.e. once per hour.
- `30 * * * *`: This is also executed once per hour, but half past each hour.
- `0 0 * * * *`: Execute the shell command once per day every midnight, i.e. at _00:00_ o'clock, or _12:00 AM_.
- `* * 1 * *`: Execute the shell command every minute, but only on the first day of each month. It's a bit silly example, but I just wanted to show that it's possible.
- `0 12 1 6 *`: Execute first of June every year at _12 PM_.
- `0 0 * * 7`: Execute every Sunday at _00:00_ / _12 AM_. Weekdays are expressed as numbers so that Sunday is `0` (or `7` if you wish) and Monday is `1`, etc.

The [Unix cron format page](https://www.ibm.com/docs/en/db2oc?topic=task-unix-cron-format) constains even more information, e.g. _ranges_ and _lists_.

## Random feature possibilities
- We could even have six values, [if _seconds_ would be added as an option, too, like is done in `node-cron` JavaScript library](https://github.com/kelektiv/node-cron#cron-ranges).
- Another library with the same name [`node-cron`](https://github.com/node-cron/node-cron) offers a validation method that can be used to check that the user defined cron expression is syntactically correct. This is important. This library also supports seconds, like the one above. (I added this one on 2022-08-18).

P.S.S. Nice that you posted to _Q & A_ 👍 . Now that it turned out to be something the plugin does not yet have, I'll move this discussion to _Ideas_. 🙂

**Reply by imed-ghomari** [reply]
Thank you for your thorough explanation, and for the workaround (I'll try doing it this way).

For the format, unix cron seems very powerful and covers many if not every use case I can think of, so yes to that. Even though, the first solution that came to my mind was a simple date picker 😅 .

But I'm more interested in how it is implemented. Let's say, for example (I'm talking about the future solution here), I set it up every day early in the morning (5am) and the computer is idle. When I open obsidian later in the day, will it recognize that the command did not trigger, and trigger it at the earliest time obsidian is active ?

**Reply by Taitava** [reply]
> Thank you for your thorough explanation, and for the workaround (I'll try doing it this way).

🙂👍 


> When I open obsidian later in the day, will it recognize that the command did not trigger, and trigger it at the earliest time obsidian is active ?

Hmm, that's a really good question! I haven't thought about it. I think implementing something like that would require the following:
- When Obsidian quits, store the current date and time during exit. (Or: Store the last time when the event was triggered - this should work as well).
- When Obsidian starts, check the time range between quitting and starting: does the scheduled time fall between the time range? If yes, then execute the shell command at startup, otherwise do not execute.
- Need to find a cron library that supports this kind of time range checking.

This feature would be optional and default to off. So, by default shell commands would only be executed if the scheduled time happens when Obsidian is running. I might develop this feature later than the initial version of the schedule event, but even the initial version should pick a library that supports this feature.

**Reply by imed-ghomari** [reply]
> I might develop this feature later than the initial version of the schedule event, but even the initial version should pick a library that supports this feature.

Great. Good luck then.
